File: server-shared/src/main/java/rikka/shizuku/server/api/RemoteProcessHolder.java
package rikka.shizuku.server.api;

import android.os.IBinder;
import android.os.ParcelFileDescriptor;
import android.os.RemoteException;

import java.util.concurrent.TimeUnit;

import eu.darken.porter.core.ServerProcess;
import moe.shizuku.server.IRemoteProcess;

public class RemoteProcessHolder extends IRemoteProcess.Stub {

    private final ServerProcess process;

    public RemoteProcessHolder(ServerProcess process) {
        this.process = process;
    }

    public RemoteProcessHolder(Process process, IBinder token) {
        this(new ServerProcess(process, token));
    }

    @Override
    public ParcelFileDescriptor getOutputStream() {
        return process.getOutputStream();
    }

    @Override
    public ParcelFileDescriptor getInputStream() {
        return process.getInputStream();
    }

    @Override
    public ParcelFileDescriptor getErrorStream() {
        return process.getErrorStream();
    }

    @Override
    public int waitFor() {
        return process.waitFor();
    }

    @Override
    public int exitValue() {
        return process.exitValue();
    }

    @Override
    public void destroy() {
        process.destroy();
    }

    @Override
    public boolean alive() throws RemoteException {
        return process.alive();
    }

    @Override
    public boolean waitForTimeout(long timeout, String unitName) throws RemoteException {
        // Parsed here so an unknown name is still an IllegalArgumentException across the binder.
        return process.waitForTimeout(timeout, TimeUnit.valueOf(unitName));
    }
}

File: server-shared/src/main/java/rikka/shizuku/server/api/RemoteProcessHolder.kt
package rikka.shizuku.server.api

import android.os.IBinder
import android.os.ParcelFileDescriptor
import android.os.RemoteException
import eu.darken.porter.core.ServerProcess
import java.util.concurrent.TimeUnit
import moe.shizuku.server.IRemoteProcess

class RemoteProcessHolder(private val process: ServerProcess) : IRemoteProcess.Stub() {

    constructor(process: Process, token: IBinder?) : this(ServerProcess(process, token))

    override fun getOutputStream(): ParcelFileDescriptor = process.getOutputStream()

    override fun getInputStream(): ParcelFileDescriptor = process.getInputStream()

    override fun getErrorStream(): ParcelFileDescriptor = process.getErrorStream()

    override fun waitFor(): Int = process.waitFor()

    override fun exitValue(): Int = process.exitValue()

    override fun destroy() {
        process.destroy()
    }

    @Throws(RemoteException::class)
    override fun alive(): Boolean = process.alive()

    @Throws(RemoteException::class)
    override fun waitForTimeout(timeout: Long, unitName: String): Boolean =
        // Parsed here so an unknown name is still an IllegalArgumentException across the binder.
        process.waitForTimeout(timeout, TimeUnit.valueOf(unitName))
}
